pkg/filesystem/local_directory_windows.go

```diff
@@ -0,0 +1,1105 @@
+// +build windows
+
+package filesystem
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/buildbarn/bb-storage/pkg/filesystem/path"
+
+	"golang.org/x/sys/windows"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	modntdll = windows.NewLazySystemDLL("ntdll.dll")
+
+	procNtSetInformationFile = modntdll.NewProc("NtSetInformationFile")
+)
+
+func ntSetInformationFile(handle windows.Handle, iosb *windows.IO_STATUS_BLOCK, fileInfo *byte, len uint32, class uint32) error {
+	r0, _, _ := syscall.Syscall6(procNtSetInformationFile.Addr(), 5, uintptr(handle),
+		uintptr(unsafe.Pointer(iosb)), uintptr(unsafe.Pointer(fileInfo)), uintptr(len), uintptr(class), 0)
+	if r0 != 0 {
+		return windows.NTStatus(r0)
+	}
+	return nil
+}
+
+const (
+	// these constants are not present in x/sys/windows
+	FILE_READ_ATTRIBUTES  = 0x80
+	FILE_READ_DATA        = 1
+	FILE_READ_EA          = 8
+	FILE_WRITE_ATTRIBUTES = 0x100
+	FILE_WRITE_DATA       = 2
+	FILE_WRITE_EA         = 0x10
+	FILE_GENERIC_READ     = windows.STANDARD_RIGHTS_READ|windows.SYNCHRONIZE|FILE_READ_ATTRIBUTES|FILE_READ_DATA|FILE_READ_EA
+	FILE_GENERIC_WRITE    = windows.STANDARD_RIGHTS_WRITE|windows.SYNCHRONIZE|FILE_WRITE_ATTRIBUTES|FILE_WRITE_DATA|FILE_WRITE_EA
+
+	SYMLINK_FLAG_RELATIVE        = 1
+	FSCTL_SET_REPARSE_POINT      = 0x900A4
+	FSCTL_QUERY_ALLOCATED_RANGES = 0x940CF
+	FSCTL_SET_SPARSE             = 0x900C4
+
+	FILE_RENAME_REPLACE_IF_EXISTS = 1
+	FILE_RENAME_POSIX_SEMANTICS   = 2
+
+	FILE_DISPOSITION_DELETE          = 1
+	FILE_DISPOSITION_POSIX_SEMANTICS = 2
+
+	FileRenameInformation        = 10
+	FileLinkInformation          = 11
+	FileDispositionInformationEx = 64
+)
+
+func ntCreateFile(handle *windows.Handle, access uint32, root windows.Handle, path string, disposition uint32, options uint32) error {
+	objectName, err := windows.NewNTUnicodeString(path)
+	if err != nil {
+		return err
+	}
+	oa := &windows.OBJECT_ATTRIBUTES{
+		Length:             0,
+		RootDirectory:      root,
+		ObjectName:         objectName,
+		Attributes:         0,
+		SecurityDescriptor: nil,
+		SecurityQoS:        nil,
+	}
+	oa.Length = uint32(unsafe.Sizeof(*oa))
+	var iosb windows.IO_STATUS_BLOCK
+	var allocSize int64 = 0
+	ntstatus := windows.NtCreateFile(handle, access, oa, &iosb, &allocSize, 0,
+		windows.FILE_SHARE_READ|windows.FILE_SHARE_WRITE|windows.FILE_SHARE_DELETE,
+		disposition, options, 0, 0)
+	if ntstatus != nil {
+		switch ntstatus.(windows.NTStatus) {
+		case windows.STATUS_NOT_A_DIRECTORY:
+			return syscall.ENOTDIR
+		case windows.STATUS_FILE_IS_A_DIRECTORY:
+			return syscall.EISDIR
+		default:
+			return ntstatus.(windows.NTStatus).Errno()
+		}
+	}
+	return nil
+}
+
+type FILE_ATTRIBUTE_TAG_INFO struct {
```

Comment:
Could we make types like these private? Or if the goal is to get all of these added to x/sys/windows at some point, move them into a separate package for the time being? pkg/filesystem/windowsext?